### PR TITLE
Update bufcheck.RunnerProvider to use the plugin config

### PR DESCRIPTION
**Pull request by emcfarlane**
Update `bufcheck.RunnerProvider` to take the `bufconfig.PluginConfig` as the argument for creating the runner. This allows for setting the runner not based on the the local path config. A helper has been provided to create local runners for a command.Runner.

### CODE REVIEW

File: private/pkg/pluginrpcutil/pluginrpcutil.go [hunk 1]
```diff
@@ -23,29 +23,3 @@ import (
...
-// RunnerProviderFunc is a function that implements RunnerProvider.
-type RunnerProviderFunc func(programName string, programArgs ...string) pluginrpc.Runner
-
-// NewRunner implements RunnerProvider.
-func (r RunnerProviderFunc) NewRunner(programName string, programArgs ...string) pluginrpc.Runner {
-	return r(programName, programArgs...)
-}
-
-// NewRunnerProvider returns a new RunnerProvider for the command.Runner.
-func NewRunnerProvider(delegate command.Runner) RunnerProvider {
```
Comment (emcfarlane) [inline private/pkg/pluginrpcutil/pluginrpcutil.go:41]: To avoid import of `bufconfig` in a `pkg` dir this has been moved to `bufcheck`.

File: private/bufpkg/bufcheck/client.go [hunk 2]
```diff
@@ -331,21 +331,17 @@ func (c *client) getMultiClient(
 		)
 	}
 	for _, pluginConfig := range pluginConfigs {
-		if pluginConfig.Type() != bufconfig.PluginConfigTypeLocal {
-			return nil, syserror.New("we only handle local plugins for now with lint and breaking")
```
Comment (emcfarlane) [inline private/bufpkg/bufcheck/client.go:335]: Error message is moved to the runner provider.

File: private/bufpkg/bufcheck/multi_client.go [hunk 3]
```diff
@@ -101,13 +101,13 @@ func (c *multiClient) Check(ctx context.Context, request check.Request) ([]*anno
...
 					return fmt.Errorf("plugin %q failed: %w", delegate.PluginName, err)
 				}
-				lock.Lock()
 				annotations := slicesext.Map(
 					delegateResponse.Annotations(),
 					func(checkAnnotation check.Annotation) *annotation {
 						return newAnnotation(checkAnnotation, delegate.PluginName)
 					},
 				)
+				lock.Lock()
```
Comment (emcfarlane) [inline private/bufpkg/bufcheck/multi_client.go:110]: Unrelated change, looks like we can move the lock to wrap just the append.

File: private/bufpkg/bufcheck/bufcheck.go [hunk 4]
```diff
@@ -162,17 +164,35 @@ func WithPluginsEnabled() ClientFunctionOption {
 	return pluginsEnabledOption{}
 }
 
-// RunnerProvider provides pluginrpc.Runners for program names and args.
+// RunnerProvider provides pluginrpc.Runners for a given plugin config.
 type RunnerProvider interface {
-	NewRunner(programName string, programArgs ...string) pluginrpc.Runner
+	NewRunner(pluginConfig bufconfig.PluginConfig) (pluginrpc.Runner, error)
```
Comment (mfridman) [inline private/bufpkg/bufcheck/bufcheck.go:169]: I believe this is the primary motivation of this PR, otherwise, this interface is tightly coupled to executing plugins locally.

By taking a `PluginConfig,` we can in the future have constructors for both `NewLocalPluginConfig` and `NewRemotePluginConfig`. This effectively punts the `pluginConfig.Type() != bufconfig.PluginConfigTypeLocal` check to the provider. The provider now has more information about what the runner should be.

**Review by bufdev — APPROVED** [review]
